[ARM] `az bicep`: Reflecting bicep cli updates to `az bicep` command

polatengin opened the pull request:
**Related command**

Following list of commands are added;

- `az bicep build-params`

- `az bicep generate-params`

- `az bicep decompile-params`

**Description**

_"bicep cli"_ got new command additions recently;

- `az bicep build-params`: builds the given `.bicepparam` file and creates '.parameters.json` file

- `az bicep generate-params`: generates a `.bicepparam` or `.parameters.json` file, based on the `--output-param {bicepparam, file}` argument, from the given `.bicep` file

- `az bicep decompile-params`: generates a `.bicepparam` file from the given `.parameters.json` file

**History Notes**
[ARM] `az bicep build-params`: Support generating `parameters.json` file from the given `bicepparam` file with the `--file` argument
[ARM] `az bicep decompile-params`: Support generating `parameters.bicepparam` file from the given `parameters.json` file with the `--file` argument
[ARM] `az bicep generate-params`: Support generating `main.parameters.json` with the parameters that doesn't have default values in the given `.bicep` file
[ARM] `az bicep generate-params`: Add new parameter `--output-format` to support generating parameter file in `bicepparam` and `json` formats
[ARM] `az bicep generate-params`: Add new parameter `--include-params` to support generating parameter file with all the parameters in the given `bicep` file, or with only parameters that doesn't have default values in the given `bicep` file

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

---

`azdev style resource` command output;

![image](https://github.com/Azure/azure-cli/assets/118744/6d4ae200-b7d2-44fc-9e16-8e485abb3769)

`azdev linter resource` command output;

![image](https://github.com/Azure/azure-cli/assets/118744/aeea4a21-74c1-4d89-947d-a4332749d3b2)

yonzhan commented:
<!-- add-pr-comment:add-pr-comment -->

BICEP

anthony-c-martin commented:
@zhoxing-ms please could we have a review on this?

shenglol reviewed (approved):
Changes look good to me, but you may need to fix the style and linter issues.

polatengin commented:
> Changes look good to me, but you may need to fix the style and linter issues.

thanks for the approval @shenglol, I just updated the PR with linter/style fixes 👍

zhoxing-ms commented:
@polatengin Please resolve these CLI linter issues~

zhoxing-ms commented:
Please refer to the [guideline](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests) to write the description of those PR changes in the `History Notes` section of PR description. Please note that this content will be synced to the [release notes](https://learn.microsoft.com/en-us/cli/azure/release-notes-azure-cli?toc=%2Fcli%2Fazure%2Ftoc.json&bc=%2Fcli%2Fazure%2Fbreadcrumb%2Ftoc.json) of CLI. Such as #26261 
![image](https://github.com/Azure/azure-cli/assets/16612065/a802c6f7-12f8-426b-a895-c50b52d12352)

zhoxing-ms commented:
Please note that we are launching the release for this sprint this week. Please resolve all comments by tomorrow, otherwise the release of this PR will have to be postponed to the next sprint **(on 09-05)**

shenglol commented:
@polatengin The CI failed because there is no tests for the new commands:
![image](https://github.com/Azure/azure-cli/assets/16367959/8f681ce7-ea4a-4d70-95ed-c011448112e1)

Can you add some tests in https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/resource/tests/latest/test_resource_custom.py?

zhoxing-ms commented:
> To reflect recent changes on the "bicep cli" to az bicep command

You need to add some more specific descriptions for these new commands and parameters so that customers can understand these new features

polatengin commented:
> Please note that we are launching the release for this sprint this week. Please resolve all comments by tomorrow, otherwise the release of this PR will have to be postponed to the next sprint **(on 09-05)**

I added tests to the PR, all the pipelines are successfully run now 👍

I also updated the PR description

Can you review this PR, please?

/cc: @zhoxing-ms @yonzhan

zhoxing-ms commented:
@polatengin Please note that the parameters do not need to be written in the second command signature section, but should be written in the following description section. And the bicep module belongs to `[ARM]` instead of `[Bicep]`
For example:
> [Bicep] `az bicep build-params --file {parameters.bicepparam}` : Generates parameters.json file

Should be

> [ARM] `az bicep build-params`: Add new parameter `--file` to support generating parameters.json file

polatengin commented:
> @polatengin Please note that the parameters do not need to be written in the second command signature section, but should be written in the following description section. And the bicep module belongs to `[ARM]` instead of `[Bicep]` For example:
> 
> > [Bicep] `az bicep build-params --file {parameters.bicepparam}` : Generates parameters.json file
> 
> Should be
> 
> > [ARM] `az bicep build-params`: Add new parameter `--file` to support generating parameters.json file

I just updated the description of this PR, can you check please @zhoxing-ms ?

zhoxing-ms commented on src/azure-cli/azure/cli/command_modules/resource/_help.py:
```diff
@@ -2817,6 +2829,22 @@
 
+helps['bicep decompile-params'] = """
+type: command
+short-summary: Attempts to decompile a parameters .json file to .bicepparam.
```
Please use the first person voice in the help message

polatengin replied:
got it! 👍

zhoxing-ms commented on src/azure-cli/azure/cli/command_modules/resource/_help.py:
```diff
@@ -2817,6 +2829,22 @@
 
+helps['bicep decompile-params'] = """
+type: command
+short-summary: Attempts to decompile a parameters .json file to .bicepparam.
```
```suggestion
short-summary: Decompile a parameters .json file to .bicepparam.
```

zhoxing-ms commented on src/azure-cli/azure/cli/command_modules/resource/_params.py:
```diff
@@ -845,6 +863,8 @@ def load_arguments(self, _):
         c.argument('outfile', arg_type=bicep_outfile_type)
         c.argument('stdout', arg_type=bicep_stdout_type)
         c.argument('no_restore', arg_type=bicep_no_restore_type, help="When set, generates the parameters file without restoring external modules.")
+        c.argument('output_format', arg_type=CLIArgumentType(options_list=['--output-format'], help="Set output format. Valid values are ( json | bicepparam )."))
+        c.argument('include_params', arg_type=CLIArgumentType(options_list=['--include-params'], help="Set include params. Valid values are ( all | required-only )."))
```
Please do not explicitly define the default option name repeatedly
```suggestion
        c.argument('output_format', help="Set output format. Valid values are ( json | bicepparam )."))
        c.argument('include_params', help="Set include params. Valid values are ( all | required-only )."))
```

polatengin commented:
/azp run

yonzhan commented:
/azp run

polatengin commented:
@zhoxing-ms , @yonzhan I see all the checks are passing 🥳

can you review, please?

zhoxing-ms commented on src/azure-cli/azure/cli/command_modules/resource/_params.py:
```diff
@@ -820,6 +827,17 @@ def load_arguments(self, _):
 
+    with self.argument_context('bicep decompile-params') as c:
+        c.argument('file', arg_type=bicep_file_type, help="The path to the parameters file to build in the file system.")
+        c.argument('bicep_file', arg_type=CLIArgumentType(options_list=['--bicep-file'],
+                                                          completer=FilesCompleter(),
+                                                          type=file_type,
+                                                          help="Path to the bicep template file (relative to the .bicepparam file) that will be referenced in the using declaration."))
```
We usually define `CLIArgumentType` when the code needs to be reused
```suggestion
        c.argument('bicep_file', completer=FilesCompleter(), type=file_type, help="Path to the bicep template file (relative to the .bicepparam file) that will be referenced in the using declaration.")
```

zhoxing-ms commented:
@polatengin Please note that we are launching the release for this sprint this week. Please resolve all comments by tomorrow, otherwise the release of this PR will have to be postponed to the next sprint (on 09-26)

polatengin commented:
> @polatengin Please note that we are launching the release for this sprint this week. Please resolve all comments by tomorrow, otherwise the release of this PR will have to be postponed to the next sprint **(on 09-26)**

hey @zhoxing-ms, I resolved all the comments. can you please check the PR and help me to merge it?

bebound commented:
This PR adds two `ScenarioTest`, which runs the recording test in Test Deb/RPM Package and fails in `tests/latest/test_resource.py::BicepDecompileParamsTest::test_bicep_decompile_params_file`.

On AMD64, it raises `Couldn't find a valid ICU package installed on the system. Please install libicu (or icu-libs) using your package manager and try again`.
We run the test in docker, this package is not installed by default. (The ADO Ubuntu 20.04 agent installs libicu-dev by default.)

On ARM64, it raises `OSError: [Errno 8] Exec format error: '/root/.azure/bin/bicep'`. `az bicep` is compatible with ARM64 arch, it download the amd64 bicep on arm64 platform.

A quick fix is to change these test to live only.
